File: src/observe.cc
#include <nan.h>
#include <stdio.h>
#include <stdlib.h>

#include "process-observe.h"


void GetChildren(const Nan::FunctionCallbackInfo<v8::Value>& info) {
  v8::Local<v8::Object> obj = Nan::New<v8::Object>();
  double pid = info[0]->NumberValue();

  int pcount;
  uint32_t* plist;
  int err = process_children(pid, &plist, &pcount);
  if (err) {
    Nan::ThrowError("Error excuting getting children");
  }

  v8::Local<v8::Number> process_count = Nan::New(pcount);
  v8::Local<v8::Array> process_arr = Nan::New<v8::Array>(pcount);

  for (size_t i = 0; i < (size_t)pcount; i++) {
    process_arr->Set(i, Nan::New<v8::Number>(plist[i]));
  }

  obj->Set(Nan::New("pids").ToLocalChecked(), process_arr);
  obj->Set(Nan::New("count").ToLocalChecked(), process_count);

  info.GetReturnValue().Set(obj);
}

void Info(const Nan::FunctionCallbackInfo<v8::Value>& info) {
  v8::Local<v8::Object> obj = Nan::New<v8::Object>();
  double pid = info[0]->NumberValue();

  p_obsrv_info_t data;

  int err = process_info(pid, &data);
  if (err) {
    Nan::ThrowError("Error excuting getting children");
  }

  obj->Set(Nan::New("pid").ToLocalChecked(), Nan::New<v8::Number>(data.pid));
  obj->Set(Nan::New("parent_pid").ToLocalChecked(), Nan::New<v8::Number>(data.ppid));
  obj->Set(Nan::New("children_count").ToLocalChecked(), Nan::New<v8::Number>(data.children_count));


  v8::Local<v8::Array> process_arr = Nan::New<v8::Array>(data.children_count);
  for (size_t i = 0; i < (size_t)data.children_count; i++) {
    process_arr->Set(i, Nan::New<v8::Number>(data.children[i]));
  }
  obj->Set(Nan::New("children").ToLocalChecked(), process_arr);

  info.GetReturnValue().Set(obj);
}

void Init(v8::Local<v8::Object> exports) {
  exports->Set(Nan::New("children").ToLocalChecked(),
               Nan::New<v8::FunctionTemplate>(GetChildren)->GetFunction());
  exports->Set(Nan::New("info").ToLocalChecked(),
               Nan::New<v8::FunctionTemplate>(Info)->GetFunction());
}

NODE_MODULE(addon, Init)

File: lib/observe.js
'use strict'
const bindings = require('./bindings')
const Readable = require('stream').Readable

/**
 * @class
 */
class ProcessObserve extends Readable {
  constructor() {
    super()
  }
}

/**
 * Query a process associated with a specified `pid` for the number of and a
 * list of pids of childnre it has.
 * @example
 * const observe = require('process-observer')
 * // query the OS pid for all its children (100+), here with sync API
 * observe.children(0)
 * // -> {pids: [1, 2, 50, ...], count: 121}
 *
 * // inspect your current process (here assuming that you have just spawned
 * // a new process), here with async API
 * observe.children(process.pid, (err, result) => {
 *   if (err)
 *     console.log(err)
 *   console.log(result)
 *   // -> {pids: [5841], count: 1}
 * })
 *
 * @param  {Number}   pid `pid` to query
 * @param  {Function} [cb] optional callback. If provided sync function will be called on next tick.
 * @return {Object}       returns an object with [0] the list of pids and [1] their count
 */
ProcessObserve.prototype.children = (pid, cb) => {
  if (typeof pid !== 'number') {
    const err = new TypeError('First argument must be a PID (number).')
    if (arguments.length === 1) {
      throw err
    } else {
      return cb(err)
    }
  }

  if (arguments.length === 2) {
    process.nextTick(() => {
      return cb(null, bindings.children(pid))
    })
  } else {
    return bindings.children(pid)
  }
};

/**
 * Query a process associated with a specified `pid` a wider process specific data
 * such as a its parent, children, hardware usage etc. (see list TBD)
 * @example
 * const observe = require('process-observer')
 * const currenProcess = process.pid
 * // sync
 * observe.info(currenProcess)
 * // -> { pid: 3804, parent_pid: 3803, children_count: 1, children: [ 3805 ]}
 *
 * // async
 * observe.info(currenProcess, (err, result) => {
 *   if (err)
 *     console.log(err)
 *   console.log(result)
 *    // -> {pid: 3804, parent_pid: 3803, children_count: 1, children: [ 3805 ]}
 * })
 *
 * @param  {Number}   pid `pid` to query
 * @param  {Function} [cb] optional callback. If provided sync function will be called on next tick.
 * @return {Object}       returns an object with process specific data
 */
ProcessObserve.prototype.info = (pid, cb) => {
  if (typeof pid !== 'number') {
    const err = new TypeError('First argument must be a PID (number).')
    if (arguments.length === 1) {
      throw err
    } else {
      return cb(err)
    }
  }

  if (arguments.length === 2) {
    process.nextTick(() => {
      return cb(null, bindings.info(pid))
    })
  } else {
    return bindings.info(pid)
  }
};

module.exports = new ProcessObserve()
